src/mlpack/methods/ann/layer/softshrink.hpp

```diff
@@ -0,0 +1,127 @@
+/**
+ * @file softshrink.hpp
+ * @author Lakshya Ojha
+ *
+ * The soft shrink function has threshold proportional to the noise level given
+ * by the user.
+ * The use of a Soft Shrink activation function provides adaptive denoising at
+ * various noise levels using a single CNN(Convolution Neural) without a
+ * requirement to train a unique CNN for each noise level.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_ANN_LAYER_SOFTSHRINK_HPP
+#define MLPACK_METHODS_ANN_LAYER_SOFTSHRINK_HPP
+
+#include <mlpack/prereqs.hpp>
+
+namespace mlpack {
+namespace ann /** Artifical Neural Network. */ {
+
+/**
+ * Soft Shrink operator is defined as,
+ * @f{eqnarray*}{
+ * f(x) &=& \left\{
+ *   \begin{array}{lr}
+ *     x - lambda & : x >  lambda \\
+ *     x + lambda & : x < -lambda \\
+ *     0 & : otherwise
+ *   \end{array} \\
+ * \right.
+ * f'(x) &=& \left\{
+ *   \begin{array}{lr}
+ *     1 & : x >  lambda \\
+ *     1 & : x < -lambda \\
+ *     0 & : otherwise
+ *   \end{array}
+ * \right.
+ * @f}
+ * lambda is set to 0.5 by default.
+ */
+template <
+    typename InputDataType = arma::mat,
+    typename OutputDataType = arma::mat
+>
+class SoftShrink
+{
+ public:
+  /**
+   * Create Soft Shrink object using specified hyperparameter lambda.
+   *
+   * @param lambda The noise level of an image depends on settings of an
+   *        imaging device. The settings can be used to select appropriate
+   *        parameters for denoising methods. It is proportional to the noise
+   *        level entered by the user.
+   *        And it is calculated by multiplying the
+   *        noise level sigma of the input(noisy image) and a
+   *        coefficient 'a' which is one of the training parameters.
+   *        Default value of lambda is 0.5.
+   */
+  SoftShrink(const double lambda = 0.5);
+
+  /**
+   * Ordinary feed forward pass of a neural network, evaluating the function
+   * f(x) by propagating the activity forward through f.
+   * 
+   * @param input Input data used for evaluating the Soft Shrink function.
+   * @param output Resulting output activation
+   */
+  template<typename InputType, typename OutputType>
+  void Forward(const InputType&& input, OutputType&& output);
```

Comment:
I think you have to refactor this according to #2259 ?

Reply:
Hi @favre49 
I think I did rebase this branch when @rcurtin asked me to do according to the same.
Thanks.

Reply:
In #2259 the method signatures of the layer functions changed to remove rvalue references, and your code conforms to the old method signature. I think you'll have to change that, correct me if I'm wrong.

Reply:
Ohh! My bad I really missed that, obviously rebase won't help with my own code.
Thanks for pointing out, will change the same ASAP.
Thanks.

Reply:
@favre49 I have implemented the above asked changes, can you please have a look at the same when you get time.
Thanks.